docs/proposals/002-api-proposal/proposal.md

```diff
@@ -0,0 +1,313 @@
+        // In the face of resource-scarcity. Higher priority requests will be preserved, and lower priority class requests will be rejected.
+        Objective *Objective
+        // Optional.
+	// Allow multiple versions of a model for traffic splitting. 
+	// If not specified, the target model name is defaulted to the modelName parameter.
+        TargetModels []common.TargetModel
+}
+
+
+
+// TargetModel represents a deployed model or a LoRA adapter. The
+// TargetModelName is expected to match the name of the LoRA adapter
+// (or base model) as it is registered within the model server. Inference
+// Gateway assumes that the model exists on the model server and is the
+// responsibility of the user to validate a correct match. Should a model fail
+// to exist at request time, the error is processed by the Instance Gateway,
+// and then emitted on the appropriate ModelGroup object.
+type TargetModel struct {
+        // The name of the adapter as expected by the ModelServer.
+        TargetModelName string
+        // Weight is used to determine the percentage of traffic that should be 
+        // sent to this target model when multiple versions of the model are specified.
+        Weight int
+}
+
+// Objective captures the latency SLO of a LLM use case.
+// In MVP, meeting the SLO is on a best effort basis.
+// Future: Extend the API for different behaviors of meeting the SLO.
+// The gateway will perform best-effort load balancing, and work with other components (e.g., autoscaler) to meet the
+// objectives. 
+type Objective struct {
+        // The AverageLatencyPerOutputToken is calculated as the e2e request latency divided by output token 
+        // length. Note that this is different from what is known as TPOT (time per output token) which only 
+        // takes decode time into account.
+        // The P95 is calculated over a fixed time window defined at the operator level.
+        DesiredAveragePerOutputTokenLatencyAtP95OverMultipleRequests
+    *time.Duration
+}
+```
+
+### Yaml Examples
+
+#### BackendPool(s)
+Here we create 2 BEPs that subscribe to services to collect the appropriate pods
+```yaml
+apiVersion: inference.x-k8s.io/v1alpha1
+kind: BackendPool
+metadata:
+  name: llama-2-pool
+  services: 
+  - llama-2-vllm
+---
+apiVersion: inference.x-k8s.io/v1alpha1
+kind: BackendPool
+metadata:
+  name: gemini-pool
+  services: 
+  - gemini-jetstream-tpu-v5e
+  - gemini-vllm-a100
+```
+
+#### Model Group
+
+Here we consume both pools with a single ModelGroup, while also specifying 2 useCases. Where `sql-code-assist` is both the name of the ModelUseCase, and the name of the LoRA adapter on the model server. And `npc-bot` has a layer of indirection for those names, as well as a specified objective.
+```yaml
+apiVersion: inference.x-k8s.io/v1alpha1
+kind: ModelGroup
+metadata:
+  name: my-model-group
+spec:
+  useCases:
```

Comment:
this ModelGroup has two useCases and BEP, with this, what is the mapping for the useCases and BEP? Can one useCase use two BEP?

Reply:
By definition, BackendPools referenced here has to hold the same useCases. It means that `sql-code-assist` is an available LoRA adapter on both BackendPool. Therefore, routing a request for `sql-code-assist` to any of the BackendPool works.

Reply:
That's correct. While routing to the BEP happens earlier(at the K8s Gateway). Allowing a user to specify all the UseCases they have, and all the BEPs they would like this model available on in a single place seemed a valuable UX, comments on the ModelGroup API definition further up goes into more detail. PTAL

Reply:
Got it, thanks @Joffref and @kfswain , just want to confirm, this means for both `sql-code-assist` and `npc-bot` are available LoRA adapter on both BackendPool, right? 

If so, shall we update the description for this YAML a little bit like the following?

Here we consume both pools with a single ModelGroup, while also specifying 2 useCases. Where `sql-code-assist` is both the name of the ModelUseCase, and the name of the LoRA adapter on the model server. And `npc-bot` has a layer of indirection for those names, as well as a specified objective. Both `sql-code-assist` and `npc-bot` are available LoRA adapter on both BackendPool and routing to the BackendPool happens earlier(at the K8s Gateway).

Reply:
>  just want to confirm, this means for both sql-code-assist and npc-bot are available LoRA adapter on both BackendPool, right? 

Yes and no, `sql-code-assist` is the name of the LoRA adapter (we know this because there is no additional configuration in the `targetModels` object, so the assumed default is the `modelName`

However for `npc-bot`, since the `targetModels` have 2 entries `npc-bot-v1` & `npc-bot-v2`, those 2 entries are the names of the lora adapter on the model server, and when a request comes in for `npc-bot` we split traffic between those 2. (Useful for a controlled rollout of a new adapter, or something like that).

Maybe that part was already understood, so forgive me if we are reviewing the same content again. But tried to reword this in the yaml description. Lemme know if this is more helpful. Thanks!

Reply:
this is super clear, thanks!